### PR TITLE
MLOps pipeline: initial working version

### PR DESCRIPTION
**Pull request by opnuub**
• added Dockerfile and cloudbuild_yaml added to run cloud build scripts and containerize codebase 
• added download_from_gcs, save_training_plot, upload_to_gcs in helpers.py 
• modified word_segmenter.py to save history plots and upload model to cloud storage if trained on vertex AI 
• added requirements.txt
• added train.py for training on the cloud
• fixed typo on README.md
• added Cloud Usage.md to walkthrough setting up ML training pipeline

### CODE REVIEW

File: lstm_word_segmentation/helpers.py [hunk 1]
```diff
@@ -84,3 +88,74 @@ def print_grapheme_clusters(thrsh, language, exclusive):
     print(ratios)
     print("number of different grapheme clusters in {} = {}".format(language, len(ratios.keys())))
     print("{} grapheme clusters form {} of the text".format(cnt, thrsh))
+
+
+def download_from_gcs(gcs_uri: str, dir: str) -> None:
+    """
+    This function downloads data from Google Cloud Storage. Given gcs_uri, the function downloads 
+    all subfolder and files inside and stores them in dir.
```
Comment (SahandFarhoodi) [inline lstm_word_segmentation/helpers.py:96]: nit: subfolder -> subfolders

File: lstm_word_segmentation/helpers.py [hunk 2]
```diff
@@ -84,3 +88,74 @@ def print_grapheme_clusters(thrsh, language, exclusive):
...
+        if rel_path == "":
+            continue
+        local_path = os.path.join(dir, rel_path)
+        os.makedirs(os.path.dirname(local_path), exist_ok=True)
+        blob.download_to_filename(local_path)
+
+def save_training_plot(history, model_path: Path) -> None:
+    """
+    Plot accuracy & loss curves from a Keras History object and save the figure.
+    Args
```
Comment (SahandFarhoodi) [inline lstm_word_segmentation/helpers.py:121]: nit: missing ":" after Args.

File: lstm_word_segmentation/helpers.py [hunk 3]
```diff
@@ -84,3 +88,74 @@ def print_grapheme_clusters(thrsh, language, exclusive):
...
+        history: The History returned by `model.fit(...)`.
+        model_path: Destination directory, e.g. Path(__file__).parent.parent.absolute().joinpath("Models", model_name)
+    """
+    fig, (ax_acc, ax_loss) = plt.subplots(1, 2, figsize=(10, 4), sharex=True)
+
+    ax_acc.plot(history.history["accuracy"],     label="train acc")
```
Comment (SahandFarhoodi) [inline lstm_word_segmentation/helpers.py:127]: Maybe call it train accuracy on the plot as well? "accuracy" can be confusing.

File: lstm_word_segmentation/helpers.py [hunk 4]
```diff
@@ -84,3 +88,74 @@ def print_grapheme_clusters(thrsh, language, exclusive):
...
+    ax_acc.plot(history.history["val_accuracy"], label="val acc")
+    ax_acc.set_title("Accuracy")
+    ax_acc.set_xlabel("Epoch")
+    ax_acc.set_ylabel("Accuracy")
+    ax_acc.legend()
+    ax_acc.grid(True)
+
+    ax_loss.plot(history.history["loss"],     label="train loss", linestyle="--")
```
Comment (SahandFarhoodi) [inline lstm_word_segmentation/helpers.py:135]: Similarly, probably better if you call it train loss.

File: train.py [hunk 5]
```diff
@@ -0,0 +1,34 @@
+from lstm_word_segmentation.word_segmenter import WordSegmenter
+from lstm_word_segmentation.helpers import download_from_gcs
+import argparse
+
+def parser_args():
+    parser = argparse.ArgumentParser()
+    parser.add_argument('--path', help = 'Dataset file on Google Cloud Storage', type = str)
+    parser.add_argument('--language', help = 'Dataset language: Thai/ Burmese', type = str, default = "Thai")
```
Comment (SahandFarhoodi) [inline train.py:8]: In the help argument, we can potentially have other languages too. e.g. we have Lao and Khmr now and in future we may have other languages. Maybe make it more like "such as Thai or Burmese". Also minor, but there shouldn't be a space after "/".

File: train.py [hunk 6]
```diff
@@ -0,0 +1,34 @@
...
+
+def main(args):
+    download_from_gcs(args['path'], 'Data')
+    if args['language'] == 'Thai':
+        word_segmenter = WordSegmenter(input_name=args['name'], input_n=50, input_t=10000, input_clusters_num=350,
+                                    input_embedding_dim=16, input_hunits=23, input_dropout_rate=0.2, input_output_dim=4,
+                                    input_epochs=args['epochs'], input_training_data="BEST",
+                                    input_evaluation_data="BEST", input_language='Thai',
+                                    input_embedding_type="codepoints")
+    else:
```
Comment (SahandFarhoodi) [inline train.py:24]: Instead of this "else" I think we should have "elif args['language'] = 'Burmese'" here and then have an else statement at the end that basically prints "unsupported language" or sth of that sort.

File: train_thai.py [hunk 7]
```diff
@@ -33,4 +33,4 @@
                                                                 word_segmenter.hunits))
 # word_segmenter.save_model()
 word_segmenter.test_model_line_by_line(verbose=True, fast=True)
-# '''
+# '''
```
Comment (SahandFarhoodi) [inline train_thai.py:36]: I don't understand this line (maybe it is me using GitHub after years). It seems the removed and added lines are identical?

File: Cloud Usage.md [hunk 8]
```diff
@@ -0,0 +1,60 @@
+To bind the repository to Google Cloud Platform, follow the steps below:
+
+1. In Google Cloud API & Services, enable 
+    - Cloud Logging API
+    - Artifact Registry API
+    - Vertex AI API
+    - Cloud Pub/Sub API
+    - Cloud Build API
+
+2. Create a Google Cloud Storage Bucket, upload the dataset into gs://bucket_name/Data/ with the following directory structure:\
```
Comment (SahandFarhoodi) [inline Cloud Usage.md:10]: nit: remove "\" from the end of the line.

**Review by SahandFarhoodi — APPROVED** [review]
Looks good to me. Thanks!

**Comment by opnuub** [comment]
@sffc @SahandFarhoodi  Hi Shane and Sahand, this is the new polished code of both MLOps pipeling and CNN model. I've also included a sample CNN model in the Models directory if anyone is interested to test it. I could also remove it if its unnecessary for now.

File: lstm_word_segmentation/word_segmenter_cnn.py [hunk 9]
```diff
@@ -0,0 +1,596 @@
...
+from .accuracy import Accuracy
+from .line import Line
+from .bies import Bies
+from .grapheme_cluster import GraphemeCluster
+from .code_point import CodePoint
+
+class WordSegmenterCNN:
+    """
+    A class that let you make a CNN, train it, and test it.
+    Args:
```
Comment (SahandFarhoodi) [inline lstm_word_segmentation/word_segmenter_cnn.py:24]: There are args you are setting in __init__ that are not defined here (everything after embedding_type). Pls add them too.

File: lstm_word_segmentation/word_segmenter_cnn.py [hunk 10]
```diff
@@ -0,0 +1,596 @@
...
+        Args:
+            input_line: the unsegmented line
+        """
+        # Finding word breakpoints
+        # Note that it is possible that input is segmented manually instead of icu. However, for both cases we set that
+        # input_type equal to "icu_segmented" because that doesn't affect performance of this function. This way we
+        # won't need unnecessary if/else for "man_segmented" and "icu_segmented" throughout rest of this function.
+        line = Line(input_line, "icu_segmented")
+
+        # x_data and y_data will be code point based if self.embedding_type is codepoints
```
Comment (SahandFarhoodi) [inline lstm_word_segmentation/word_segmenter_cnn.py:301]: Here and in other places, you have "code point". I think they all need to be "codepoint".
